# Анализ базы данных сервиса для чтения книг по подписке

Компания купила крупный сервис для чтения книг по подписке. 

Задача — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Чтение файла

In [1]:
import pandas as pd
from sqlalchemy import create_engine 

In [2]:
# устанавливаем параметры
db_config = {'user': '', # имя пользователя
    'pwd': '', # пароль
    'host': '',
    'port': 0, # порт подключения
    'db': ''} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
     db_config['pwd'],
     db_config['host'],
     db_config['port'],
     db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

def sql(query):
    return pd.io.sql.read_sql(query, con = engine)

In [3]:
sql('''
    SELECT *
    FROM books
    LIMIT 10
''')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [4]:
sql('''
    SELECT *
    FROM reviews
    LIMIT 10
''')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

In [5]:
sql('''
    SELECT *
    FROM authors
    LIMIT 10
''')

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [6]:
sql('''
    SELECT *
    FROM publishers
    LIMIT 10
''')

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [7]:
sql('''
    SELECT *
    FROM ratings
    LIMIT 10
''')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

- Посчитайте, сколько книг вышло после 1 января 2000 года;

In [8]:
sql('''
    SELECT COUNT(*) AS books_count
    FROM books
    WHERE publication_date > '2000-01-01'
''')

books_count
0          819

В базе есть 819 книг, которые были опубликованы с 2000 года, при этом всего - 1000 книг.

- Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [9]:
sql('''
    SELECT b.book_id, 
           b.title,
           br.review_count,
           rb.avg_rating
    FROM books AS b
    LEFT JOIN (
        SELECT rev.book_id,
               COUNT(rev.review_id) AS review_count
        FROM reviews AS rev
        GROUP BY rev.book_id
    ) AS br ON b.book_id = br.book_id
    LEFT JOIN ( 
        SELECT rat.book_id,
               AVG(rat.rating) AS avg_rating
        FROM ratings as rat
        GROUP BY rat.book_id
    ) AS rb ON b.book_id = rb.book_id
    ORDER BY book_id
''')

book_id                                              title  review_count  \
0          1                                       'Salem's Lot           2.0   
1          2                 1 000 Places to See Before You Die           1.0   
2          3  13 Little Blue Envelopes (Little Blue Envelope...           3.0   
3          4  1491: New Revelations of the Americas Before C...           2.0   
4          5                                               1776           4.0   
..       ...                                                ...           ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)           3.0   
996      997                        Xenocide (Ender's Saga  #3)           3.0   
997      998                                    Year of Wonders           4.0   
998      999                        You Suck (A Love Story  #2)           2.0   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...           4.0   

     avg_rating  
0      3.666667  
1      2.500000  
2      4.666667  
3      4.500000  
4      4.000000  
..          ...  
995    3.666667  
996    3.400000  
997    3.200000  
998    4.500000  
999    3.833333  

[1000 rows x 4 columns]

Посчитали кол-во обзоров и среднюю оценку для каждой книги, есть те, у которых нет оценок и отзывов.

- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [10]:
sql('''
    SELECT p.publisher_id,
           p.publisher,
           COUNT(b.book_id) AS books_count
    FROM publishers as p
    JOIN books AS b ON b.publisher_id = p.publisher_id
    WHERE b.num_pages > 50 
    GROUP BY p.publisher_id
    ORDER BY books_count DESC
    limit 1
''')

publisher_id      publisher  books_count
0           212  Penguin Books           42

Издательство, которое выпустило наибольшее число книг (42 шт. из базы) толще 50 стр. - Penguin Books.

- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [11]:
sql('''
    SELECT a.author_id, a.author, AVG(b.agv_book_rating) as avg_rating
    FROM authors as a
    JOIN (
        SELECT b.book_id, b.author_id, COUNT(*) as ratings_count, AVG(r.rating) as agv_book_rating
        FROM ratings as r
        JOIN books as b ON r.book_id = b.book_id
        GROUP BY b.book_id, b.author_id
        ORDER BY b.book_id
    ) as b ON a.author_id = b.author_id
    WHERE ratings_count >= 50
    GROUP BY a.author_id
    ORDER BY avg_rating DESC
    LIMIT 1
''')

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.283844

В этой базе самым популярным автором является Джоан Роулинг.

- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [12]:
sql('''
    SELECT AVG(r.review_count) as avg_rating
    FROM (
        SELECT COUNT(rev.review_id) AS review_count
        FROM reviews as rev
        WHERE rev.username IN (
                            SELECT rat.username
                            FROM ratings AS rat
                            GROUP BY rat.username
                            HAVING COUNT(rat.rating_id) > 50
        )
        GROUP BY rev.username
    ) AS r
''')

avg_rating
0   24.333333

Пользователи, которые поставили более 50 оценок, написали в среднем ~24 обзора.

## Вывод

- Всего в базе данных 1000 книг, из них 819 - книги, опубликованные с 2000 года.
- Есть книги без отзывов
- Издательство, которое выпустило наибольшее число книг (42 шт. из базы) толще 50 стр. - Penguin Books. Мало книг толщиной менее 50 стр.
- В этой базе самым популярным писателем является Джоан Роулинг.
- Пользователи, которые поставили более 50 оценок, написали в среднем ~24 обзора.